### PR TITLE
Add Win Numa test-cases.

### PR DESCRIPTION
**Pull request by sarathnandu**
### Description 
Add an initial version of test-case for testing Win Numa behavior.

Fixes # - _issue number(s) if exists_

- [x] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [ ] new feature - _change that adds functionality_
- [x] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [x ] added - _required for new features and some bug fixes_
- [ ] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [x] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
_List users with `@` to send notifications_

### CODE REVIEW

File: test/tbb/test_numa_dist.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            // dummy work
+            double& b = tls_dummy.local();
+            for (int k = 0; k < 1000000; ++k)
+                b += pow(i * k, 1. / 3.);
+
+            PROCESSOR_NUMBER proc;
+            if (GetThreadIdealProcessorEx(GetCurrentThread(), &proc))
+            {
+                tls.local() = std::pair<int, int>(proc.Group, proc.Number);
```
Comment (vossmjp) [inline test/tbb/test_numa_dist.cpp:92]: I think you should check if tls.local() is already set. If so, its .first should also be equal to proc.Group. Since you are only keeping the last value in tls.local, the test would pass if a thread were to execute a chunk of iterations from the wrong processor group before moving to the correct one for the final chunk it executes. While I doubt that would happen, I think the test could be slightly modified to test for that case anyway.

File: test/tbb/test_numa_dist.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            }
+        });
+    for (const auto it : tls) {
+		validateProcgrp[it.first]++;
+	}
+	REQUIRE(validateProcgrp[master_thread_proc_grp] == requested_parallelism);
```
Comment (vossmjp) [inline test/tbb/test_numa_dist.cpp:98]: Another unlikely corner case is that the master_thread_proc_grp has the requested_parallelism count but there are counts for other processor groups too.  All other groups should have zeros but that is not checked by the test.

File: test/tbb/test_numa_dist.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    numa() : processorGroupCount(GetMaximumProcessorGroupCount()), maxProcessors(GetActiveProcessorCount(ALL_PROCESSOR_GROUPS)){   
+        numaProcessors.resize(processorGroupCount);
+       for (WORD i = 0; i < processorGroupCount; i++) {
+            this->numaProcessors[i] = GetActiveProcessorCount((i));
+        }        
+    }
+};
+
+
+void TestNumaDistribution(){
```
Comment (AnuyaWelling2801) [inline test/tbb/test_numa_dist.cpp:70]: Maybe change this to 
int TestNumaDistribution(std::vector<DWORD> &validateProcgrp), so that the function can be reused

File: test/tbb/test_numa_dist.cpp (hunk 2 above)
Comment (AnuyaWelling2801) [inline test/tbb/test_numa_dist.cpp:98]: Move to TEST_CASE so that the function can be reused.

File: test/tbb/test_numa_dist.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	std::vector<DWORD> validateProcgrp;
+	validateProcgrp.resize(GetMaximumProcessorGroupCount());
+	PROCESSOR_NUMBER proc;
+	struct numa nodes;
+	GetThreadIdealProcessorEx(GetCurrentThread(), &proc);
+	int master_thread_proc_grp = proc.Group;
+	auto requested_parallelism = nodes.numaProcessors.at(master_thread_proc_grp);
```
Comment (AnuyaWelling2801) [inline test/tbb/test_numa_dist.cpp:77]: auto requested_parallelism = nodes.numaProcessors.at(master_thread_proc_grp) + additionalParallelism;
If additionalParallelism = 0 - we have a test case for threads = cores on the node
if additionalParallelism > 0 - we can check for overflow

File: test/tbb/test_numa_dist.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    tbb::enumerable_thread_specific< std::pair<int, int> > tls;
+	tbb::enumerable_thread_specific< double > tls_dummy;
+	
+    tbb::parallel_for(0, 1000, [&](int i)
+        {
+            // dummy work
+            double& b = tls_dummy.local();
+            for (int k = 0; k < 1000000; ++k)
+                b += pow(i * k, 1. / 3.);
+
```
Comment (dnmokhov) [inline test/tbb/test_numa_dist.cpp:88]: It is probably not a good idea to assume that every thread will be engaged because of this. Something more reliable, e.g., a barrier might be better.

**Review by dnmokhov — CHANGES_REQUESTED** [review]
Please resolve the CI issues:

1. Copyright.
2. If this will stay a Windows-only test, we should only add it there. Otherwise, a Linux equivalent is needed.

File: test/tbb/test_numa_dist.cpp (hunk 1 above)
Comment (AnuyaWelling2801) [reply test/tbb/test_numa_dist.cpp]: Do we still need this if we are using a static partitioner and a barrier?

File: test/tbb/test_numa_dist.cpp (hunk 2 above)
Comment (AnuyaWelling2801) [reply test/tbb/test_numa_dist.cpp]: Fixed.

File: test/tbb/test_numa_dist.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+#include "common/spin_barrier.h"
+#include "common/exception_handling.h"
+#include "common/concepts_common.h"
+#include "test_partitioner.h"
+
+#include <cstddef>
+#include <vector>
+
+//! \file test_parallel_for.cpp
+//! \brief Test for [algorithms.parallel_for] specification
```
[the remark this answers is not in the document]
Comment (AnuyaWelling2801) [reply test/tbb/test_numa_dist.cpp]: Fixed.

File: test/tbb/test_numa_dist.cpp (hunk 5 above)
Comment (AnuyaWelling2801) [reply test/tbb/test_numa_dist.cpp]: Fixed.

File: test/tbb/test_numa_dist.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
+/*
+    Copyright (c) 2005-2021 Intel Corporation
```
Comment (dnmokhov) [inline test/tbb/test_numa_dist.cpp:2]: update

File: test/tbb/test_numa_dist.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+TEST_CASE("Numa overflow") {
+    numa example;
+    std::vector<DWORD> validateProcgrp;
+
+    int numaGrp = TestNumaDistribution(validateProcgrp, 1, 0);
+    std::vector<DWORD> result(GetMaximumProcessorGroupCount(), 0);
+    result[numaGrp] = example.numaProcessors[numaGrp];
+    result[(numaGrp+1)% GetMaximumProcessorGroupCount()] = 1;
+    REQUIRE(validateProcgrp == result);
```
Comment (dnmokhov) [inline test/tbb/test_numa_dist.cpp:135]: This is failing CI. For 1 NUMA node, `additionalParallelism=1` results in `maxProcessors+1` total threads.

**Review by dnmokhov — CHANGES_REQUESTED** [review]

File: test/tbb/test_numa_dist.cpp (hunk 8 above)
Comment (sarathnandu) [reply test/tbb/test_numa_dist.cpp:133]: Fixed in commit 3.

File: test/tbb/test_numa_dist.cpp (hunk 7 above)
Comment (AnuyaWelling2801) [reply test/tbb/test_numa_dist.cpp]: Fixed

**Comment by phprus** [comment]
``SYNCHRONIZATION_BARRIER`` required "Windows 8 / Windows Server 2012" but all other WinAPI calls required "Windows 7 / Windows Server 2008 R2".

Maybe wrap the test in an ``#if _WIN32_WINNT >= 0x0602`` check or reuse existing ``SpinBarrier`` (https://github.com/oneapi-src/oneTBB/blob/801df0edeb6ed19283744f1e4a41653ed58e7513/test/common/spin_barrier.h#L88)?

**Comment by phprus** [comment]
Thanks @AnuyaWelling2801!
All this branch tests run successfully on Windows 7.

**Review by dnmokhov — APPROVED** [review]
